### PR TITLE
HDDS-7137. Add CLI for Getting the failed deleted block txn

### PR DESCRIPTION
**Pull request by Xushaohong**
## What changes were proposed in this pull request?

Supplement CLI for deleted block txn.

  1. Integrate the related CLI
  2. Add new CLI to list txns
  3. Support JSON format file as output and input
  4. Improve UT

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7137

## How was this patch tested?
UT
WX20220818-165155
WX20220818-165230
WX20220818-165132
WX20220818-164911

### CODE REVIEW

**Comment by Xushaohong** [comment]
Moving the ```DeletedBlocksTransaction``` to common ```hdds.proto``` is an ideal solution, 
but it will lead to the compatibility problem, here adds the import of  ```ScmServerDatanodeHeartbeatProtocol.proto```.

**Comment by Xushaohong** [comment]
@adoroszlai @ChenSammi PTAL~

**Comment by ChenSammi** [comment]
Hey @Xushaohong , could you past some example output of this command? For both console output and file output.

**Comment by Xushaohong** [comment]
> Hey @Xushaohong , could you past some example output of this command? For both console output and file output.

@ChenSammi  Sure,  I screenshotted three pics, and the result of `reset` is not actually worked due to the non-instant flush as you know

show3
show2
show1

**Comment by errose28** [comment]
> Moving the DeletedBlocksTransaction to common hdds.proto is an ideal solution,
but it will lead to the compatibility problem, so I adds the import of ScmServerDatanodeHeartbeatProtocol.proto

I don't think we want this dependency chain. hdds.proto exists to prevent the admin and heartbeat proto from depending on each other. Cross client and rolling upgrade compatability (future work) will be harder to maintain with this dependency. The DeletedBlocksTransaction message is not very complicated. Could we make a different version of that message to be used in the client response placed in ScmAdminProtocol.proto?

**Comment by Xushaohong** [comment]
> > Moving the DeletedBlocksTransaction to common hdds.proto is an ideal solution,
> > but it will lead to the compatibility problem, so I adds the import of ScmServerDatanodeHeartbeatProtocol.proto
> 
> I don't think we want this dependency chain. hdds.proto exists to prevent the admin and heartbeat proto from depending on each other. Cross client and rolling upgrade compatability (future work) will be harder to maintain with this dependency. The DeletedBlocksTransaction message is not very complicated. Could we make a different version of that message to be used in the client response placed in ScmAdminProtocol.proto?

The problem mentioned is right,  as the migration of proto would bring the compatibility issue,  I just considered the reuse form. I have now added the `DeletedBlocksTransactionInfo`

**Comment by errose28** [comment]
Thanks for the proto update @Xushaohong. I probably won't have time to review the rest of this PR but that part looks good.

**Comment by Xushaohong** [comment]
> Thanks for the proto update @Xushaohong. I probably won't have time to review the rest of this PR but that part looks good. Just a question: I see the new proto is in hdds.proto. Would it make more sense to be in SCMAdminProtocol.proto since it is used as the response to an SCM admin command?

Hi @errose28, hdds.proto is a common client proto,  my motivation for putting the definition here has two reasons:
1. The `DeletedBlocksTransactionInfo` could be used for the future client request, it is a prospective placement. Refer to the proto of `ContainerInfoProto` and `Pipeline`.
2. I think the placement in  ScmServerDatanodeHeartbeatProtocol.proto is a historical problem. Originally it is only considered to be used in HB, not in client queries. Such common proto shall be unified and could help reduce confusion for the user and reader. So I wonder if someday we change the reference of this TXN proto from the ScmServerDatanodeHeartbeatProtocol.proto to the hdds.proto. And we could deprecate the old reference.

Pls feel free to correct me if i am wrong~

**Comment by errose28** [comment]
> I think the placement in ScmServerDatanodeHeartbeatProtocol.proto is a historical problem. Originally it is only considered to be used in HB, not in client queries. Such common proto shall be unified and could help reduce confusion for the user and reader. So I wonder if someday we change the reference of this TXN proto from the ScmServerDatanodeHeartbeatProtocol.proto to the hdds.proto. And we could deprecate the old reference.

I think this is a good point. If we put the new message in ScmAdminProtocol.proto we will have the two messages forever since we need to support all old clients. If we put the new message in hdds.proto then later we could switch datanode and SCM communication to use the new message in hdds.proto instead of the old one in ScmServerDatanodeHeartbeatProtocol.proto  since we do not support rolling upgrades yet.

**Comment by swamirishi** [comment]
@errose28 Can this be merged?

**Comment by kerneltime** [comment]
@Xushaohong can you please rebase and resolve conflicts?

**Comment by Xushaohong** [comment]
> @Xushaohong can you please rebase and resolve conflicts?

@kerneltime  done

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto [hunk 1]
```diff
@@ -451,3 +451,10 @@ message ContainerBalancerConfigurationProto {
     required bool shouldRun = 18;
     optional int32 nextIterationIndex = 19;
 }
+
+message DeletedBlocksTransactionInfo {
+    required int64 txID = 1;
```
Comment (kerneltime) [inline hadoop-hdds/interface-client/src/main/proto/hdds.proto:456]: In general, we should let the proto layer mark args as optional and deal with it on the server side? I would mark most of these args as optional. cc @errose28

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java [hunk 2]
```diff
@@ -191,7 +200,7 @@ public int resetCount(List<Long> txIDs) throws IOException, TimeoutException {
     lock.lock();
     try {
       if (txIDs == null || txIDs.isEmpty()) {
-        txIDs = getFailedTransactions().stream()
+        txIDs = getFailedTransactions(-1).stream()
```
Comment (kerneltime) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java:203]: I would avoid magic numbers, you can define some constants for these which make it obvious what is intended Example: `LIST_ALL_FAILED_TRANSACTIONS`

**Comment by kerneltime** [comment]
@Xushaohong this looks like a pretty straightforward change, please address the minor nits.

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto (hunk 1 above)
Comment (Xushaohong) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto:456]: @kerneltime I have a little confusion here, this ```DeletedBlocksTransactionInfo``` is returned by the server, and the client now would never construct it as request args.

Comment (Xushaohong) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto]: Got it, I have updated it. @kerneltime

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java [hunk 3]
```diff
@@ -126,18 +126,27 @@ public DeletedBlockLogImpl(ConfigurationSource conf,
...
     try {
       final List<DeletedBlocksTransaction> failedTXs = Lists.newArrayList();
       try (TableIterator<Long,
           ? extends Table.KeyValue<Long, DeletedBlocksTransaction>> iter =
                deletedBlockLogStateManager.getReadOnlyIterator()) {
-        while (iter.hasNext()) {
-          DeletedBlocksTransaction delTX = iter.next().getValue();
-          if (delTX.getCount() == -1) {
-            failedTXs.add(delTX);
+        if (count < 0) {
```
[the remark this answers is not in the document]
Comment (Xushaohong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java]: done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java [hunk 4]
```diff
@@ -126,18 +128,27 @@ public DeletedBlockLogImpl(ConfigurationSource conf,
...
+          while (iter.hasNext()) {
+            DeletedBlocksTransaction delTX = iter.next().getValue();
+            if (delTX.getCount() == -1) {
+              failedTXs.add(delTX);
+            }
+          }
+        } else {
+          while (iter.hasNext() && failedTXs.size() < count) {
+            DeletedBlocksTransaction delTX = iter.next().getValue();
+            if (delTX.getCount() == -1) {
```
Comment (kerneltime) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java:149]: Should this be switched as well?
```suggestion
            if (delTX.getCount() == LIST_ALL_FAILED_TRANSACTIONS) {
```

Comment (Xushaohong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java]: Oops, this is the special value of DeletedBlocksTransaction, not for the clear instant description.       
See `DeletedBlocksTransaction.Builder builder = block.toBuilder().setCount(-1);`

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java [hunk 5]
```diff
@@ -126,18 +128,27 @@ public DeletedBlockLogImpl(ConfigurationSource conf,
...
     try {
       final List<DeletedBlocksTransaction> failedTXs = Lists.newArrayList();
       try (TableIterator<Long,
           ? extends Table.KeyValue<Long, DeletedBlocksTransaction>> iter =
                deletedBlockLogStateManager.getReadOnlyIterator()) {
-        while (iter.hasNext()) {
-          DeletedBlocksTransaction delTX = iter.next().getValue();
-          if (delTX.getCount() == -1) {
-            failedTXs.add(delTX);
+        if (count <= 0) {
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java:139]: We should use LIST_ALL_FAILED_TRANSACTIONS check here.

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto [hunk 6]
```diff
@@ -458,4 +458,11 @@ message TransferLeadershipRequestProto {
 }
 
 message TransferLeadershipResponseProto {
+}
+
+message DeletedBlocksTransactionInfo {
+    optional int64 txID = 1;
+    optional int64 containerID = 2;
+    repeated int64 localID = 3;
+    optional int32 count = 4;
```
Comment (ChenSammi) [inline hadoop-hdds/interface-client/src/main/proto/hdds.proto:467]: @Xushaohong , there is a DeletedBlocksTransaction defined in StorageContainerDatanodeProtocolProtos, which has the same fields as this one. Could we use that directly?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/GetFailedDeletedBlocksTxnSubcommand.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+  @CommandLine.ArgGroup(multiplicity = "1")
+  private TransactionsOption group;
+
+  static class TransactionsOption {
+    @CommandLine.Option(names = {"-a", "--all"},
+        description = "Get all the failed transactions.")
+    private boolean getAll;
+
+    @CommandLine.Option(names = {"-n", "--num"},
```
Comment (ChenSammi) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/GetFailedDeletedBlocksTxnSubcommand.java:53]: "-n", "--num" -> "-c", "--count"

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java [hunk 8]
```diff
@@ -41,25 +50,45 @@ public class ResetDeletedBlockRetryCountSubcommand extends ScmSubcommand {
...
+        description = "Reset the only given deletedBlock transaction ID" +
             " list, e.g 100,101,102.(Separated by ',')")
     private List<Long> txList;
-  }
 
-  @CommandLine.ParentCommand
-  private ScmAdmin parent;
+    @CommandLine.Option(names = {"-i", "--in"},
+        description = "Use file as input, need to be JSON Array format and" +
+            " contains multi \"txID\" key.")
```
Comment (ChenSammi) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java:66]: Can you add an example of JSON file here, can be just in code comments.

File: hadoop-hdds/common/pom.xml [hunk 9]
```diff
@@ -226,6 +226,10 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
       <groupId>org.apache.ozone</groupId>
       <artifactId>hdds-interface-admin</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.ozone</groupId>
+      <artifactId>hdds-interface-server</artifactId>
+    </dependency>
```
Comment (ChenSammi) [inline hadoop-hdds/common/pom.xml:232]: @Xushaohong , which class reference requires this new dependency?

File: hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto [hunk 10]
```diff
@@ -488,6 +491,15 @@ message ReplicationManagerReportResponseProto {
   required ReplicationManagerReportProto report = 1;
 }
 
+message GetFailedDeletedBlocksTxnRequestProto {
+  optional string traceID = 1;
+  required int32 count = 2;
```
Comment (ChenSammi) [inline hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto:496]: In batch fetch support, we need a start ID and batch size, so user can do the iterate using different start ID and same batch size.  So we should add a start TxnID here.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java [hunk 11]
```diff
@@ -41,25 +50,45 @@ public class ResetDeletedBlockRetryCountSubcommand extends ScmSubcommand {
...
+      Gson gson = new Gson();
+      List<Long> txIds;
+      try (InputStream in = new FileInputStream(group.fileName);
+           Reader fileReader = new InputStreamReader(in,
+               StandardCharsets.UTF_8)) {
+        DeletedBlocksTransactionInfoWrapper[] txns = gson.fromJson(fileReader,
+            DeletedBlocksTransactionInfoWrapper[].class);
+        txIds = Arrays.stream(txns).map(DeletedBlocksTransactionInfoWrapper::
+            getTxID).distinct().collect(Collectors.toList());
+        System.out.println("Load txIDS: " + txIds);
```
Comment (ChenSammi) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java:85]: Please output the start txID, and total number of txIds here.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java (hunk 5 above)
Comment (Xushaohong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImpl.java]: Updated

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto (hunk 6 above)
Comment (Xushaohong) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto:467]: As I discussed with Ethan above, it is a historical problem that the **DeletedBlocksTransaction**  in `ScmServerDatanodeHeartbeatProtocol.proto` is not in `hdds.proto` as a general PB, here if we want to reuse it we need to import `ScmServerDatanodeHeartbeatProtocol.proto` in `ScmAdminProtocol.proto` which may cause some possible compatibility issue in the future, and hence not a good option.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/GetFailedDeletedBlocksTxnSubcommand.java (hunk 7 above)
Comment (Xushaohong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/GetFailedDeletedBlocksTxnSubcommand.java]: done

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java (hunk 8 above)
Comment (Xushaohong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java]: image
Updated as above

File: hadoop-hdds/common/pom.xml (hunk 9 above)
Comment (Xushaohong) [reply hadoop-hdds/common/pom.xml]: Originally it is for DeletedBlocksTransaction in **DeletedBlocksTransactionInfoWrapper**,  I have now moved the wrapper class  to a better place.

File: hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto (hunk 10 above)
Comment (Xushaohong) [reply hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto:496]: Updated

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java (hunk 11 above)
Comment (Xushaohong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/ResetDeletedBlockRetryCountSubcommand.java:85]: image

**Comment by Xushaohong** [comment]
image
image
The latest usage is as above

File: hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto [hunk 12]
```diff
@@ -488,6 +491,16 @@ message ReplicationManagerReportResponseProto {
...
+  optional int32 startTxId = 3;
```
Comment (ChenSammi) [inline hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto:497]: Tx ID is int64, and long in java code.

Comment (Xushaohong) [reply hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto]: Good catch! Updated

**Comment by ChenSammi** [comment]
The last patch LGTM + 1. 
Thanks @Xushaohong for the contribution and @errose28 @swamirishi @kerneltime for the code review.

**Pull request merged**